pymc3/ode/ode.py

```diff
@@ -0,0 +1,170 @@
+
+        #Private
+        self._n = n_states
+        self._m = n_odeparams + n_states
+
+        self._augmented_times = np.insert(times, t0, 0)
+        self._augmented_func = augment_system(func, self._n, self._m)
+        self._sens_ic = self._make_sens_ic()
+
+        self._cached_y = None
+        self._cached_sens = None
+        self._cached_parameters = None
+
+    def _make_sens_ic(self):
+        # The sensitivity matrix will always have consistent form.
+        # If the first n_odeparams entries of the parameters vector in the simulate call
+        # correspond to ode paramaters, then the first n_odeparams columns in
+        # the sensitivity matrix will be 0
+        sens_matrix = np.zeros((self._n, self._m))
+
+        # If the last n_states entrues of the paramters vector in the simulate call
+        # correspond to initial conditions of the system,
+        # then the last n_states columns of the sensitivity matrix should form
+        # an identity matrix
+        sens_matrix[:, -self.n_states:] = np.eye(self.n_states)
+
+        # We need the sensitivity matrix to be a vector (see augmented_function)
+        # Ravel and return
+        dydp = sens_matrix.ravel()
+
+        return dydp
+
+    def _system(self, Y, t, p):
+        '''
+        This is the function that will be passed to odeint.
+        Solves both ODE and sensitivities
+        Args:
+            Y (vector): current state and current gradient state
+            t (scalar): current time
+            p (vector): parameters
+        Returns:
+            derivatives (vector): derivatives of state and gradient
+        '''
+
+        dydt, ddt_dydp = self._augmented_func(Y[:self._n], t, p, Y[self._n:])
+        derivatives = np.concatenate([dydt, ddt_dydp])
+        return derivatives
+
+    def _simulate(self, parameters):
+        # Initial condition comprised of state initial conditions and raveled
+        # sensitivity matrix
+        y0 = np.concatenate([ parameters[self.n_odeparams:] , self._sens_ic])
+
+        # perform the integration
+        sol = scipy.integrate.odeint(func=self._system,
+                                     y0=y0,
+                                     t=self._augmented_times,
+                                     args=tuple([parameters]))
+        # The solution
+        y = sol[1:, :self.n_states]
+
+        # The sensitivities, reshaped to be a sequence of matrices
+        sens = sol[1:, self.n_states:].reshape(len(self.times), self._n, self._m)
+
+        return y, sens
+
+    def _cached_simulate(self, parameters):
+        if np.array_equal(np.array(parameters), self._cached_parameters):
+            return self._cached_y, self._cached_sens
+        else:
+            return self._simulate(np.array(parameters))
+
+    def state(self, x):
+        y, sens = self._cached_simulate(np.array(x, dtype=np.float64))
+        self._cached_y, self._cached_sens, self._cached_parameters = y, sens, x
+        return y.ravel()
```

Comment:
This is a breaking change.  Here is the traceback when I try to rerun the notebook.
```

TypeError                                 Traceback (most recent call last)
<ipython-input-3-16b9d87cbe1d> in <module>
     13     #If we know one of the parameter values, we can simply pass the value.
     14     #No need to specify a prior.
---> 15     ode_solution = ode_model(odeparams = [gamma, 9.8], y0 = [0]).reshape(yobs.shape)
     16 
     17     Y = pm.Normal('Y', mu = ode_solution, sd = sigma, observed = yobs)

~/anaconda3/envs/gsoc/lib/python3.6/site-packages/theano/tensor/var.py in reshape(self, shape, ndim)
    319                                  str(type(ndim)))
    320 
--> 321         return theano.tensor.basic.reshape(self, shape, ndim=ndim)
    322 
    323     def dimshuffle(self, *pattern):

~/anaconda3/envs/gsoc/lib/python3.6/site-packages/theano/tensor/basic.py in reshape(x, newshape, ndim)
   5064                 "argument to 'reshape' to avoid this problem." % newshape)
   5065     op = Reshape(ndim)
-> 5066     rval = op(x, newshape)
   5067     return rval
   5068 

~/anaconda3/envs/gsoc/lib/python3.6/site-packages/theano/gof/op.py in __call__(self, *inputs, **kwargs)
    623             for i, ins in enumerate(node.inputs):
    624                 try:
--> 625                     storage_map[ins] = [self._get_test_value(ins)]
    626                     compute_map[ins] = [True]
    627                 except AttributeError:

~/anaconda3/envs/gsoc/lib/python3.6/site-packages/theano/gof/op.py in _get_test_value(cls, v)
    560             # ensure that the test value is correct
    561             try:
--> 562                 ret = v.type.filter(v.tag.test_value)
    563             except Exception as e:
    564                 # Better error message.

~/anaconda3/envs/gsoc/lib/python3.6/site-packages/theano/tensor/type.py in filter(self, data, strict, allow_downcast)
    176             raise TypeError("Wrong number of dimensions: expected %s,"
    177                             " got %s with shape %s." % (self.ndim, data.ndim,
--> 178                                                         data.shape))
    179         if not data.flags.aligned:
    180             try:

TypeError: For compute_test_value, one input test value does not have the requested type.
 
Backtrace when that variable is created:

  File "/Users/demetri/anaconda3/envs/gsoc/lib/python3.6/site-packages/IPython/core/interactiveshell.py", line 2881, in _run_cell
    return runner(coro)
  File "/Users/demetri/anaconda3/envs/gsoc/lib/python3.6/site-packages/IPython/core/async_helpers.py", line 68, in _pseudo_sync_runner
    coro.send(None)
  File "/Users/demetri/anaconda3/envs/gsoc/lib/python3.6/site-packages/IPython/core/interactiveshell.py", line 3058, in run_cell_async
    interactivity=interactivity, compiler=compiler, result=result)
  File "/Users/demetri/anaconda3/envs/gsoc/lib/python3.6/site-packages/IPython/core/interactiveshell.py", line 3249, in run_ast_nodes
    if (await self.run_code(code, result,  async_=asy)):
  File "/Users/demetri/anaconda3/envs/gsoc/lib/python3.6/site-packages/IPython/core/interactiveshell.py", line 3326, in run_code
    exec(code_obj, self.user_global_ns, self.user_ns)
  File "<ipython-input-3-16b9d87cbe1d>", line 15, in <module>
    ode_solution = ode_model(odeparams = [gamma, 9.8], y0 = [0]).reshape(yobs.shape)
  File "/Users/demetri/anaconda3/envs/gsoc/lib/python3.6/site-packages/theano/gof/op.py", line 615, in __call__
    node = self.make_node(*inputs, **kwargs)
  File "/Users/demetri/anaconda3/envs/gsoc/lib/python3.6/site-packages/pymc3/ode/ode.py", line 155, in make_node
    return theano.Apply(self, [x], [x.type()])

The error when converting the test value to that variable type:
Wrong number of dimensions: expected 1, got 2 with shape (20, 1).
```

Reply:
Is this after you removed the ravel? There's still the reshape that should be be removable..

Reply:
Yes, this is after making the suggested change.
No dice.  Removing the `.reshape(yobs.shape)` results in the same error.

Reply:
This could be the reason why I had `infer_shape` in the other `Op`...